Codes/Visualization/ConstructOrderBookVisualization_20200214.R

```diff
@@ -45,41 +61,95 @@ mydata$clocktime<-format(as.POSIXct((mydata$time),origin="1970-01-01",tz="UTC"),
+      #discard orders that have been completely executed/cancelled by timestamp
+      i<-match(subt$orderID,totalect$orderID)
+      subt$execanc<-totalect$size[i]; rm(i) #this is the amount of the submitted order that has been executed or cancelled as of timestamp
+      subt$execanc<-replace(subt$execanc,is.na(subt$execanc)==1,0) #replace missing with 0
+      subt$remainder<-subt$size-subt$execanc #this is the amount of the order remaining as of the timestamp
+      if (length(q<-which(subt$remainder<0))!=0){stop("Negative remainders")} #sanity check
+      rm(execanct,totalect)
+      }
+      else {subt$remainder<-subt$size}
 subt<-subt[subt$remainder!=0,] #these should represent the orders that have not yet been execulated
-rm(execanct,totalect)
+#discard buy orders that are higher than the best bid and sell orders that are lower than the best ask
+subt<-subt[which((subt$direction==1&subt$price<=orbk$bidprice1)|(subt$direction==-1&subt$price>=orbk$askprice1)),]
 
 #aggregate across MPID type, price and direction
 subt$temp<-paste(subt$mpid,subt$price,subt$direction,sep=";")
-temp<-aggregate(subt$size,by=list(Category=subt$temp),FUN=sum)
-
-s<-t(matrix(unlist(strsplit(temp$Category,split=";")),3,nrow(temp)))
-temp$Category<-NULL
-temp$mpid<-s[,1]
-temp$price<-as.numeric(s[,2])
-temp$direction<-as.integer(s[,3])
+suba<-aggregate(subt$size,by=list(Category=subt$temp),FUN=sum)
+s<-t(matrix(unlist(strsplit(suba$Category,split=";")),3,nrow(suba)))
+suba$Category<-NULL
+suba$mpid<-s[,1]
+suba$price<-as.numeric(s[,2])
+suba$direction<-as.integer(s[,3])
 rm(s)
+suba<-suba[order(suba$price),]
 
+#only keep x best buy and sell orders
+#x<-50
+#buy<-suba[suba$direction==1,]; buy<-buy[order(buy$price,decreasing=T),]
+#buy<-buy[1:x,]; buy<-buy[order(buy$price),]
+#sell<-suba[suba$direction==-1,]; sell<-sell[order(sell$price),]
+#sell<-sell[1:x,]; sell<-sell[order(sell$price),]
+#suba<-rbind(buy,sell)
 
-##subt[which(subt$orderID=="1059989"),]
+#subt[which(subt$orderID=="1059989"),]
 #totalect[which(totalect$orderID=="1059989"),]
 
-
+suba$interval<-y
+colnames(suba)<-cols
+subplot<-rbind(subplot,suba)
+subold<-subt[,1:48]
+rm(suba,subt)
+
+oldtesttime<-testtime
+xlabels[y]<-testtime
+testtime<-testtime+60
+
+}
+
+#xlabels<-xlabels+60 #such that they represent the ending time of the interval
+
+pdfname<-paste0(direc,stock,"_11-",day,"_",xlabels[1],"_",xlabels[length(xlabels)],".pdf")
+pdf(file=pdfname,width=6,height=4,paper="special")
+xlabels<-format(as.POSIXct((xlabels),origin="1970-01-01",tz="UTC"),"%H:%M")
+p<-ggplot(subplot,aes(x=factor(interval),y=price,color=mpid))+scale_colour_manual(values=c("null"="grey","SBSH"="green","TMBR"="blue","UBSS"="red","WEMM"="pink"))+
+  geom_point(alpha=1/4,aes(size=size))+scale_x_discrete(name="",breaks=seq(1,60,10),labels=xlabels[seq(1,60,10)])+
+  ggtitle(paste(stock,"Nov",day))
+for (i in 1:length(min)){
```

Comment:
So "+" indeed represents the mid-quote. Why the plot is so asymmetric? --  On the ask side quotes are crowded, while on the bid side they are so sparse

Reply:
I was wondering the same thing. I think it is not always so asymmetric, i.e., the book looks a bit more balanced if you look at one hour later, from 11:30-12:30pm. Maybe it says something about the direction in which traders expect the price to move?

Reply:
Interesting, isn't it? Thank you for explanation.